### ISSUE TITLE
related-list:region: промис fetchRegionList() переживает detach() и всё равно зовёт onSelect()

### ISSUE THREAD

**Issue by kalbac** [kalbac/woodev-plugin-framework#573]
Побочное наблюдение при закрытии #524. **Прочитано, не замерено** — записано именно так.

## Механика

`attachRelatedListRegion` (`woodev/shipping-method/assets/js/frontend/location-select-modes.js:135-268`):

```js
fetchRegionList().then( function( entries ) {
    for ( var i = 0; i < entries.length; i++ ) {
        if ( candidate.record.label === text ) {
            options.onSelect( { record: candidate.record } )   // <-- после возможного detach()
            releaseIfHeld()
            return
        }
    }
    releaseIfHeld()
} )
```

`detach()` у этого рендерера делает только `unbind()` — снимает слушатель `change`. Он **не отменяет промис и не ставит флаг «отсоединён»**, так что уже летящий `fetchRegionList()` доедет до `.then()` и позовёт `options.onSelect()` с записью региона.

Момент штатный, а не экзотический: `detach()` зовётся на `updated_checkout`, и там же WooCommerce перерисовывает фрагмент адреса, а каскад запускает свой `reconcileAfterCheckoutUpdate()`.

Плюс `fetchRegionList()` — это `GET /location/list`, **10,5 секунды на холодном регионе** (замерено в #541, цитируется прямо в докблоке этого же обработчика). Окно между кликом и ответом широкое.

## Чего НЕ хватает для оценки

Не замерено ни одного из двух:

1. Что живой сценарий реально даёт `detach()` во время полёта именно этого запроса.
2. Что поздний `onSelect()` действительно вредит. Возможно, и нет: `options.onSelect` принадлежит КАСКАДУ, а не рендереру, каскад после `updated_checkout` жив, и выбор региона — это факт, о котором ему как раз надо знать. Тогда текущее поведение правильное, и карточку надо закрыть решением, а не правкой.

Именно поэтому это отдельная карточка, а не довесок к #524: чинить непроверенное — против правил проекта.

## Что сделать

1. Воспроизвести: кликнуть регион, дождаться `detach()` (триггер `updated_checkout`) ДО ответа `/location/list`, посмотреть, что делает каскад с прилетевшим `onSelect()`.
2. Если вредит — гасить так же, как `activeAbort` гасит ajax в `buildSelectField()`, либо флагом «отсоединён» в обоих коллбэках.
3. Если не вредит — записать решением прямо в докблок `detach()`, чтобы следующий читатель не заводил эту карточку заново.

## Связано

- #524 — при закрытии чего найдено
- #541 — откуда замер 10,5 с
- #449 — предыдущий заход на teardown-щели в этом файле

**Comment by kalbac** [comment]
## Сверка по коду в рамках #644, 31.08.2026 (s109) — механика подтверждена, а опасность оказалась НЕ той, что в карточке

Карточка честно помечена «прочитано, не замерено». Замерил по коду обе половины. Первая
подтверждается, вторая — переворачивается.

### 1. Механика: подтверждена дословно

`attachRelatedListRegion.detach()` (`location-select-modes.js:265-267`) — это ровно `unbind()` и
ничего больше. Ни отмены запроса, ни флага «отсоединён», ни проверки в `.then()`. Для сравнения, в
этом же файле `buildSelectField()` умеет отменять свой летящий ajax (`activeAbort`, ~строка 565,
и её докблок прямо говорит, что `detach()` иначе не дотянется до текущего запроса). То есть
асимметрия внутри одного файла, а не забытая мелочь вообще.

### 2. «Вредит ли поздний onSelect» — по коду НЕ вредит по той причине, которую карточка предполагала

Проверил, куда именно приземляется поздний вызов. `onSelectFor( entry, node )`
(`location-cascade.js:3064`) замыкается на объект `entry`. Ключевой вопрос — переживает ли `entry`
перерисовку. **Переживает:**

- `entries` строится ОДИН раз при инициализации модуля (`location-cascade.js:1221`);
- `handleCheckoutUpdated()` (строка 4581) просто перебирает тот же массив;
- `reconcileAfterCheckoutUpdate( entry )` (строка 4543) получает ТОТ ЖЕ объект и внутри зовёт
  `detachOne( entry, … )` / `attachOne( entry, node )` — пересоздаётся виджет, а не состояние.

Значит поздний `onSelect()` пишет `entry.records[ level ]` в ЖИВОЕ состояние каскада, а не в
мёртвый объект. Гипотеза 2 карточки («каскад после `updated_checkout` жив, и выбор региона — это
факт, о котором ему как раз надо знать») кодом подтверждается.

### 3. Но при этом всплывает ДРУГАЯ опасность, которой в карточке нет

Опасен не `detach()`. Опасно **вытеснение**.

`reconcileAfterCheckoutUpdate()` делает `detachOne` + `attachOne` — то есть на том же узле
появляется НОВЫЙ виджет с НОВЫМ замыканием (и своим собственным `lastHandledText`, он
per-closure). Окно `/location/list` — **10,5 с на холодном регионе**, цифра из #541 и цитируется в
докблоке самого обработчика. В этом окне покупатель вполне может выбрать регион ещё раз, уже на
новом виджете. Тогда порядок такой:

1. клик A → летит запрос A;
2. `updated_checkout` → detach/attach, новый виджет;
3. клик B → летит запрос B, приходит быстрее (регион уже прогрет) → `entry.records.region = B`;
4. приходит ответ A → `entry.records.region = A`, `backwardsFill`, `refreshAddressLock`,
   `enqueueSelect( entry, A )`.

Последним побеждает **устаревший** выбор. Это уже не «поздний вызов в мёртвый объект», а
классический last-writer-wins, и лечится он тем же средством, что предлагает пункт 2 карточки, —
только обоснование другое, и флага «отсоединён» тут МАЛО: нужен токен поколения запроса, как у
`activeAbort` в соседнем рендерере.

### Что осталось замерить и почему не мной

Достижимость шага 3 — вопрос к живому ригу: нужно успеть кликнуть второй регион внутри окна
ответа. Статически я показал, что путь существует и что состояние общее; сколько это стоит на
практике — вопрос воспроизведения. **Оператора это не требует**, требует сессии с ригом.

Карточку не закрываю и не переписываю (приоритеты его, #644). Но её пункт 3 — «если не вредит,
записать решением в докблок» — теперь выбирать нельзя не глядя: по первой причине не вредит, по
второй может.

**Comment by kalbac** [comment]
## Замер (s120) — из двух вопросов карточки закрыт первый, по второму есть ответ, но не риговый

Карточка была записана честно: «прочитано, не замерено». Меряю.

### 1. Промис действительно переживает `detach()` — ЗАМЕРЕНО

Детерминированный пробник в jest на существующей оснастке
`tests/js/location-select-modes.test.js` (там `fetchJson` отдаёт промис с ручным `resolve`, так что
гонку не нужно ловить таймингом — её можно построить):

```
attach → change («Москва») → api.detach() → resolve(/location/list)
```

```
PROBE after change:               fetch calls = 1   onResolving = 1
PROBE after resolve-post-detach:  onSelect    = 1   release     = 1
```

То есть **отсоединённый рендерер зовёт и `onSelect()`, и `release()`**. Механика карточки
подтверждена ровно как описана: `detach()` делает только `unbind()`, промис не отменяется и флага
«отсоединён» нет.

### 2. `release()` вреда НЕ несёт — снимается чтением реализации

`onResolvingFor()` (`location-cascade.js:3214-3231`) возвращает `release()`, который сравнивает
токен:

```js
if ( null !== token && entry.selectBusy && entry.selectBusy.token === token ) {
    clearSelectBusy( entry );
}
```

Свой маркер он уже не найдёт, если его сменил новый выбор, новый уровень или settled `/select`.
Поздний `release()` от мёртвого рендерера просто стоит в стороне. **Этот путь чинить нечего.**

### 3. Поздний `onSelect()` вреда НЕСЁТ — и вот почему

`options.onSelect` каскада уходит в `enqueueSelect( entry, record )`
(`location-cascade.js:2071-2102`), а его первая же строка:

```js
entry.pendingRecord = record;
```

**Последний записавший выигрывает — и это не недосмотр, а заявленная гарантия функции.** Её
собственный докблок:

> GUARANTEE: once the customer stops selecting, the record persisted server-side equals their MOST
> RECENT selection

Очередь одноместная именно потому, что сервер держит один слот текущей локации на entry, и
одновременные POST'ы могут ответить не в том порядке. Защита построена на посылке, что **каждый
вызывающий — живой**: «позже вызвали» тождественно «клиент выбрал позже».

Отсоединённый рендерер эту посылку ломает. Его `onSelect()` приезжает как свежий выбор, хотя за ним
стоит клик десятисекундной давности, и перезаписывает `pendingRecord` более новым по времени, но
более старым по смыслу регионом. Очередь защититься не может: понятия «протухший» у неё нет by
design.

Достижимая последовательность: клик по региону A → `updated_checkout` (штатный) отсоединяет рендерер
→ фрагмент адреса перерисован, новый рендерер прицеплен → клиент выбирает регион B → быстрый ответ
B → **и только потом** ответ A (окно 10,5 с из #541). Победит A.

### Что из этого ЗАМЕРЕНО, а что ВЫВЕДЕНО — граница

- **Замерено:** пункт 1 (пробник выше) и контракт `enqueueSelect`/`release` (чтение реализующего
  кода, а не комментариев рядом).
- **НЕ замерено:** сама последовательность целиком на риге. Пункт 1 карточки требовал живого
  повтора; я построил его детерминированно в jest, что строже по механике, но это НЕ доказательство
  того, что боевой `updated_checkout` попадает именно в это окно на реальном чекауте.

Поэтому вывод формулирую так: **дефект настоящий по конструкции, частота не измерена.**

### Что делать

Карточка сама называет верное направление (п. 2 её плана): гасить так же, как `activeAbort` гасит
ajax в `buildSelectField()`, либо флагом «отсоединён», который проверяют ОБА коллбэка. Учитывая п. 2
выше, гвард нужен строго перед `options.onSelect()`; `releaseIfHeld()` можно оставить как есть — он
безвреден и его безвредность держится токеном, а не удачей.

Карточка остаётся открытой: замер сделан, починка не бралась.
